### PR TITLE
Adding Xray Source Control Service

### PR DESCRIPTION
**Pull request by EyalDelarea**
- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-client-go/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Adding new XSC service which built on top of Xray.
Any functionality that is not specifically implemented inside XSC will fallback to Xray.

Main changes:
- Turnning Xray manger into an interface, which now can be splitted into Xray and XSC.
**Important note**: I didn't change the Xray name as this is the product name for the entire JFrog Security services.
- Implemented the new `XSC graph scan` function to support linking scan info into repositories.

### CODE REVIEW

File: xray/manager/xray.go [hunk 1]
```diff
@@ -197,3 +187,28 @@ func (sm *XrayServicesManager) IsEntitled(featureId string) (bool, error) {
 	entitlementsService.XrayDetails = sm.config.GetServiceDetails()
 	return entitlementsService.IsEntitled(featureId)
 }
+
+// IsXscEnabled will try to get XSC version. If route is not available, user is not entitled for XSC.
+func (sm *XrayServicesManager) IsXscEnabled() (xscEntitled bool, xsxVersion string, err error) {
```
Comment (sverdlov93) [inline xray/manager/xray.go:192]:
```suggestion
func (sm *XrayServicesManager) IsXscEnabled() (xsxVersion string, err error) {
```

no need to have boolean and version response
xsc is enabled when version is not empty

File: xray/manager/xsc.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/jfrog/jfrog-client-go/http/jfroghttpclient"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/xray/scan"
+)
+
+type XscServicesManger struct {
+	XrayServicesManager
+}
+
+func (xsc *XscServicesManger) IsXscEnabled() (bool, string, error) {
```
Comment (sverdlov93) [inline xray/manager/xsc.go:13]:
```suggestion
func (xsc *XscServicesManger) IsXscEnabled() (string, error) {
```

File: xray/scan/scan.go [hunk 3]
```diff
@@ -273,6 +296,16 @@ type JfrogResearchSeverityReason struct {
...
+	GitRepoName   string   `json:"git_repo_name"`
+	GitProject    string   `json:"git_project"`
+	GitProvider   string   `json:"git_provider"`
+	Technologies  []string `json:"technologies"`
+	BranchName    string   `json:"branch_name"`
+	LastCommit    string   `json:"last_commit"`
+	CommitHash    string   `json:"commit_hash"`
+	CommitMessage string   `json:"commit_message"`
+	CommitAuthor  string   `json:"commit_author"`
+	Date          string   `json:"date"`
```
Comment (sverdlov93) [inline xray/scan/scan.go:310]: why date?

File: xray/manager/xray.go [hunk 4]
```diff
@@ -197,3 +187,28 @@ func (sm *XrayServicesManager) IsEntitled(featureId string) (bool, error) {
 	entitlementsService.XrayDetails = sm.config.GetServiceDetails()
 	return entitlementsService.IsEntitled(featureId)
 }
+
+// IsXscEnabled will try to get XSC version. If route is not available, user is not entitled for XSC.
```
Comment (sverdlov93) [inline xray/manager/xray.go:191]:
```suggestion
// IsXscEnabled will try to get XSC version. If route is not available then Xsc is not enabled.
```

File: xray/scan/xscScan.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	scanResponse := ScanResponse{}
+	if err = json.Unmarshal(body, &scanResponse); err != nil {
+		return nil, errorutils.CheckErrorf("couldn't parse JFrog Xray server response: " + err.Error())
+	}
+	if scanResponse.ScannedStatus == XrayScanStatusFailed {
+		// Failed due to an internal Xray error
+		return nil, errorutils.CheckErrorf("received a failure status from JFrog Xray server:\n%s", errorutils.GenerateErrorString(body))
+	}
+	return &scanResponse, err
+}
```
Comment (sverdlov93) [inline xray/scan/xscScan.go:135]: file name : sscScan > xscscan

File: xray/scan/xscScan.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+type XscScanService struct {
+	ScanService
+}
+
+func NewXscScanService(client *jfroghttpclient.JfrogHttpClient, details auth.ServiceDetails) *XscScanService {
+	return &XscScanService{ScanService{client: client, XrayDetails: details}}
+}
+
+func (xsc *XscScanService) SendScanContext(details *XscGitInfoContext) (multiScanId string, err error) {
```
Comment (sverdlov93) [inline xray/scan/xscScan.go:34]:
```suggestion
func (xsc *XscScanService) SendScanGitInfoContext(details *XscGitInfoContext) (multiScanId string, err error) {
```

File: xray/scan/xscScan.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	url := xsc.XrayDetails.GetXscUrl() + postScanContextAPI
+	resp, body, err := xsc.client.SendPost(url, requestBody, &httpClientsDetails)
+	if err != nil {
+		return
+	}
+	if err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK, http.StatusCreated); err != nil {
+		scanErrorJson := ScanErrorJson{}
+		if e := json.Unmarshal(body, &scanErrorJson); e == nil {
+			return "", errorutils.CheckErrorf(scanErrorJson.Error)
+		}
```
Comment (sverdlov93) [inline xray/scan/xscScan.go:54]: is it needed here?

File: xray/scan/xscScan.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+		return
+	}
+	scanResponse := XscPostContextResponse{}
+	if err = json.Unmarshal(body, &scanResponse); err != nil {
+		return "", errorutils.CheckError(err)
+	}
+	return scanResponse.MultiScanId, err
+}
+
+func (xsc *XscScanService) ScanGraph(scanParams *XrayGraphScanParams) (string, error) {
```
Comment (sverdlov93) [inline xray/scan/xscScan.go:64]: is it duplicate with Xray scanGraph?

File: tests/xsc_test.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
+package tests
```
Comment (sverdlov93) [inline tests/xsc_test.go:1]: if a get this right all this test are currently unit tests and they can sit near the xsc service code and no neet the XSC testing suite

Comment (EyalDelarea) [reply tests/xsc_test.go:1]: in the client i've put just a simple unit test to make sure things don't break

File: xray/services/xsc.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	log.Debug("Scanning graph using XSC service...")
+	scanService := NewXscScanService(xsc.client, xsc.config.GetServiceDetails())
+	multiScanId, err := scanService.SendScanGitInfoContext(params.XscGitInfoContext)
+	if err != nil {
+		// Don't fail the entire scan when failed to send XscGitInfoContext
+		log.Warn("failed to send xsc git info context with the following error: ", err.Error())
+	}
+	if multiScanId != "" {
+		params.MultiScanId = multiScanId
+	}
```
Comment (sverdlov93) [inline xray/services/xsc.go:34]: if param.gitinfo != nil 
then: 
<all this>

**Comment by EyalDelarea** [comment]
open a new pull request 
https://github.com/jfrog/jfrog-client-go/pull/823

**Pull request closed without merging**